src/java.base/share/classes/javax/crypto/KDF.java

```diff
@@ -0,0 +1,631 @@
+     *     the provider to use for this key derivation
+     *
+     * @return a {@code KDF} object
+     *
+     * @throws NoSuchAlgorithmException
+     *     if the specified provider does not support the specified KDF
+     *     algorithm
+     * @throws InvalidAlgorithmParameterException
+     *     if the {@code KDFParameters} is an invalid value
+     * @throws NullPointerException
+     *     if the {@code algorithm} or {@code provider} is {@code null}
+     */
+    public static KDF getInstance(String algorithm,
+                                  KDFParameters kdfParameters,
+                                  Provider provider)
+        throws NoSuchAlgorithmException, InvalidAlgorithmParameterException {
+        Objects.requireNonNull(algorithm, "null algorithm name");
+        Objects.requireNonNull(provider, "provider may not be null");
+        try {
+            Instance instance = GetInstance.getInstance("KDF", KDFSpi.class,
+                                                        algorithm,
+                                                        kdfParameters,
+                                                        provider);
+            if (!JceSecurity.canUseProvider(instance.provider)) {
+                String msg = "JCE cannot authenticate the provider "
+                             + instance.provider.getName();
+                throw new SecurityException(msg);
+            }
+            return new KDF((KDFSpi) instance.impl, instance.provider, algorithm,
+                           kdfParameters);
+
+        } catch (NoSuchAlgorithmException nsae) {
+            return handleException(nsae);
+        }
+    }
+
+    private static KDF handleException(NoSuchAlgorithmException e)
+        throws NoSuchAlgorithmException, InvalidAlgorithmParameterException {
+        Throwable cause = e.getCause();
+        if (cause instanceof InvalidAlgorithmParameterException) {
+            throw (InvalidAlgorithmParameterException) cause;
+        }
+        throw e;
+    }
+
+    /**
+     * Derives a key, returned as a {@code SecretKey}.
+     * <p>
+     * The {@code deriveKey} method may be called multiple times on a particular
+     * {@code KDF} instance.
+     *
+     * @param alg
+     *     the algorithm of the resultant {@code SecretKey} object
+     * @param kdfParameterSpec
+     *     the object describing the inputs to the derivation function
+     *
+     * @return a {@code SecretKey} object corresponding to a key built from the
+     *     KDF output and according to the derivation parameters. Rather than
+     *     permitting a {@code null} return value here, implementations will
+     *     handle cases where a value cannot be returned by throwing an
+     *     appropriate {@code Exception}.
+     *
+     * @throws InvalidAlgorithmParameterException
+     *     if the information contained within the {@code kdfParameterSpec} is
+     *     invalid, if {@code alg} is invalid, or if their combination
```

Comment:
Again, what is the definition of `alg` being invalid? Would it be clearer to just refer to `KDFSpi.engineDeriveKey()` and not duplicating the words here?